File: ShapeByte/features/workout/item/ui/ItemSetsState.swift
//
//  ItemSetsState.swift
//  ShapeByte
//
//  Created by Lang, Stefan [Shape Byte Tech] on 31.07.24.
//

import Foundation

enum ItemSetsState: Equatable {
    case idle
    case started(totalSets: Int)
    case running(currentSet: Int, totalSets: Int, currentSetProgress: Progress, totalProgress: Progress)
    case paused
    case finished

    // TODO: test
    var isRunning: Bool {
        let retVal: Bool

        switch self {
        case .idle:
            retVal = false
        case .started(let totalSets):
            retVal = true
        case .running(let currentSet, let totalSets, let currentSetProgress, let totalProgress):
            retVal = true
        case .paused:
            retVal = false
        case .finished:
            retVal = false
        }

        return retVal
    }

}

File: shared/src/commonMain/kotlin/de/stefan/lang/shapebyte/features/workout/domain/ItemSetsState.kt
package de.stefan.lang.shapebyte.features.workout.domain

import de.stefan.lang.shapebyte.features.workout.data.ItemSetData
import de.stefan.lang.shapebyte.utils.Progress

sealed interface ItemSetsState {
    data object Idle : ItemSetsState
    data class Started(val totalSets: Int) : ItemSetsState

    sealed interface Running : ItemSetsState {
        val currentSet: Int
        val totalSets: Int
        val setData: ItemSetData

        data class SetStarted(
            override val currentSet: Int,
            override val totalSets: Int,
            override val setData: ItemSetData,
        ) : Running

        data class SetRunning(
            override val currentSet: Int,
            override val totalSets: Int,
            val currentSetProgress: Progress,
            val totalProgress: Progress,
            override val setData: ItemSetData,
        ) : Running

        data class SetFinished(
            override val currentSet: Int,
            override val totalSets: Int,
            override val setData: ItemSetData,
        ) : Running
    }

    data object Paused : ItemSetsState
    data object Finished : ItemSetsState

    // TODO: test
    val isRunning: Boolean
        get() {
            return this is Running || this is Started
        }
}
